sphinx/ext/autodoc/__init__.py

```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
             kwargs.setdefault('show_annotation', False)
 
-        # for classes, the relevant signature is the __init__ method's
-        initmeth = self.get_attr(self.object, '__init__', None)
-        # classes without __init__ method, default __init__ or
-        # __init__ written in C?
-        if initmeth is None or \
-                inspect.is_builtin_class_method(self.object, '__init__') or \
-                not(inspect.ismethod(initmeth) or inspect.isfunction(initmeth)):
-            return None
-        try:
-            self.env.app.emit('autodoc-before-process-signature', initmeth, True)
-            sig = inspect.signature(initmeth, bound_method=True)
+        def get_user_defined(obj, attr):
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
+                return None
+            return attr
+
+        sig = None
+
+        # this sequence is copied from inspect._signature_from_callable
+
+        # First, let's see if it has an overloaded __call__ defined
+        # in its metaclass
+        if sig is None:
+            call = get_user_defined(type(self.object), '__call__')
+
+            if call is not None:
+                # types which have confusing metaclass signatures it would be best
+                # not to show
+                call_blacklist = [
+                    'enum.EnumMeta.__call__',
+                ]
+                if "{0.__module__}.{0.__qualname__}".format(call) in call_blacklist:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except TypeError:
+                    print("Faild, oh no")
+
+        # Now we check if the 'obj' class has a '__new__' method
+        if sig is None:
+            new = get_user_defined(self.object, '__new__')
+            if new is not None:
+                self.env.app.emit('autodoc-before-process-signature', new, True)
+                try:
+                    sig = inspect.signature(new, bound_method=True)
+                except TypeError:
+                    print("Faild, oh no")
+
+        # Finally, we should have at least __init__ implemented
+        if sig is None:
+            init = get_user_defined(self.object, '__init__')
+            if init is not None:
+                self.env.app.emit('autodoc-before-process-signature', init, True)
+                try:
+                    sig = inspect.signature(init, bound_method=True)
+                except TypeError:
+                    print("Faild, oh no")
+
+        # if none of the attributes are user-defined, we can skip
+        # `autodoc-before-process-signature`, and just let inspect handle it
```

Comment:
Why skipped here?

Reply:
Because if the type is native, there's no source code to process, and the assumption is `autodoc-before-process-signature` has nothing to do.

Reply:
Updated the comment a little.

Reply:
I think this event is public one. So it should be emitted even if no source code.

Reply:
The only way to ensure it is emitted would be to either:
* Emit it for the object itself on this line, rather than attempting to find the right `__new__` / `__init__` etc
* Duplicating even more of the logic inside `inspect.signature`